### PR TITLE
8322077: Add Ideal transformation: (~a) | (~b) => ~(a & b)

### PR DESCRIPTION
**Pull request by CptGit**
Hello,

(~a) | (~b) => ~(a & b) is a widely seen pattern, for example it is implemented for LLVM [here](https://github.com/llvm/llvm-project/blob/397f1ce9efb4eea1ee10fe4833f733b8c7abd878/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L1617C28-L1617C28); however it is missing in current implementation of hotspot. This pull request adds this transformation and associated tests.

Thanks.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8322077](https://bugs.openjdk.org/browse/JDK-8322077): Add Ideal transformation: (~a) | (~b) =&gt; ~(a &amp; b) (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [dc60a548](https://git.openjdk.org/jdk/pull/16334/files/dc60a5488ab770a6b65649dbcd72c4c7897d48fa)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**) ⚠️ Review applies to [b21e242b](https://git.openjdk.org/jdk/pull/16334/files/b21e242b306d9cf01e1ba84ebba47ef5b471d98e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16334/head:pull/16334` \
`$ git checkout pull/16334`

Update a local copy of the PR: \
`$ git checkout pull/16334` \
`$ git pull https://git.openjdk.org/jdk.git pull/16334/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16334`

View PR using the GUI difftool: \
`$ git pr show -t 16334`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16334.diff">https://git.openjdk.org/jdk/pull/16334.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16334#issuecomment-1856292931)

### CODE REVIEW

**Comment by CptGit** [comment]
Hi, can I get a review?

**Comment by TobiHartmann** [comment]
A JBS issue has been created for this: [JDK-8322077](https://bugs.openjdk.org/browse/JDK-8322077). Please update the PR accordingly.

**Comment by CptGit** [comment]
> A JBS issue has been created for this: [JDK-8322077](https://bugs.openjdk.org/browse/JDK-8322077). Please update the PR accordingly.

Thanks. Updated.

**Comment by merykitty** [comment]
LGTM, but may you consider having dedicated functions to check and create `Not` patterns? Thanks a lot.

**Comment by CptGit** [comment]
> LGTM, but may you consider having dedicated functions to check and create `Not` patterns? Thanks a lot.

Thanks. I updated. Please let me know if there are more comments.

**Review by TobiHartmann — COMMENTED** [review]
Looks good to me. Please update the copyright dates.

I submitted testing and will report back once it passed.

File: src/hotspot/share/opto/addnode.cpp [hunk 1]
```diff
@@ -790,6 +808,13 @@ Node* OrINode::Ideal(PhaseGVN* phase, bool can_reshape) {
       return new RotateRightNode(in(1)->in(1), shift, TypeInt::INT);
     }
   }
+
+  // Convert "~a | ~b" into "~(a & b)"
+  if (is_not(phase, in(1), T_INT) && is_not(phase, in(2), T_INT)) {
+    return make_not(phase,
+                    phase->transform(new AndINode(in(1)->in(1), in(2)->in(1))),
+                    T_INT);
```
Comment (eme64) [inline src/hotspot/share/opto/addnode.cpp:816]: I'd put the `AndI` node on a separate line. Call it `add_a_b` or similar.
Then you can transform on the next line.
And then on a third line the `make_not`.

File: src/hotspot/share/opto/addnode.cpp [hunk 2]
```diff
@@ -768,6 +768,24 @@ Node* rotate_shift(PhaseGVN* phase, Node* lshift, Node* rshift, int mask) {
...
+  switch (bt) {
+    case T_INT:
+      return new XorINode(n, phase->intcon(-1));
+    case T_LONG:
+      return new XorLNode(n, phase->longcon(-1L));
+    default:
+      fatal("Not implemented for %s", type2name(bt));
+  }
+  return nullptr;
+}
```
Comment (eme64) [inline src/hotspot/share/opto/addnode.cpp:787]: If you are going to use this also for your changes in https://github.com/openjdk/jdk/pull/16333, then you probably want this to go into a shared file.

**Review by eme64 — CHANGES_REQUESTED** [review]
Looks like a good idea. Left a few comments.

I would have merged this with https://github.com/openjdk/jdk/pull/16333, since it is essentially the symmetric case. But leave it separate now.

It would be nice to have some shared tests, where both optimizations need to be combined. Like:
`(~a | ~b) & (~c | ~d)` -> `~(a & b) & ~(c & d)` -> `~((a & b) | (c & d))`

**Comment by CptGit** [comment]
@eme64 @TobiHartmann Thanks for the comments. All addressed.

I rebased this PR onto #16333 so I was able to add these tests for using both optimizations. (the history was messed up).

**Comment by eme64** [comment]
Nice, looks much better, thanks for the updates!

File: test/hotspot/jtreg/compiler/c2/irTests/DeMorganLawIntTests.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ */
+package compiler.c2.irTests;
+
+import jdk.test.lib.Asserts;
+import compiler.lib.ir_framework.*;
+
+/*
+ * @test
+ * @bug 8322077
+ * @summary Test that Ideal transformations on the De Morgan's Law performe
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/irTests/DeMorganLawIntTests.java:31]:
```suggestion
 * @summary Test that Ideal transformations on the De Morgan's Law perform
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise.

/reviewers 2

**Review by TobiHartmann — APPROVED** [review]

**Review by eme64 — APPROVED** [review]
LGTM, and thanks for the work!

Please only integrate this once your other change is integrated, and merged into this one. Then wait for GHA to complete, and run your own testing.

**Review by TobiHartmann — APPROVED** [review]

**Comment by eme64** [comment]
@CptGit can you merge from master again, please? It looks now like you are pushing both the changes here and the ones from your previous PR. Once you did that, I'd like to run some testing before we push this.

**Comment by CptGit** [comment]
@eme64 Yes I merged. Looks clean now.

**Comment by eme64** [comment]
Testing running for commit 34...

**Comment by TobiHartmann** [comment]
All tests passed, this is good to go.

**Comment by CptGit** [comment]
@eme64 @TobiHartmann Thanks for testing! Can you sponsor it when you get a chance thank you!

**Pull request closed without merging**